File: src/services/instructions/loadstore/ST_S_MREGX_XX_REGY.java
package emulator.source.loadstore

import emulator.engine.CpuContext
 import Instruction from '../Instruction'import REGISTER_VALUE_NAME_MAPPER from '@/constants/registers'

export default class ST_S_MREGX_XX_REGY extends Instruction {
	public ST_S_MREGX_XX_REGY(memory, address, source, 
 destination) {
		super(memory, address, source, destination)
		super.setArgument32()
		super.setAssembler("st.s [" + this.sdest + " + 0x%08x], " + this.ssource)
	}
	
	exec ({ context, memory }) {
		context.memory[fix(context.getReg(this.destination)   +  this.argument) / 2] = (short) context.getReg(this.source) 

		context.pc  += 6
		updateViewer(context, fix(context.getReg(this.destination)  + this.argument), context.getReg(this.source) )
	}
}

File: src/services/instructions/loadstore/StSMRegXXXRegY.js
import Instruction from '../Instruction'
import { REGISTER_VALUE_NAME_MAPPER } from '@/constants/registers'

const ASSEMBLER_INSTRUCTION_EXPRESSION = (sdestination, ssource) => `st.s [${sdestination} + %s], ${ssource}`

export default class StSMRegXXXRegY extends Instruction {
	constructor (memory, address, source, destination, symbolTable) {
		super(memory, address, source, destination, symbolTable)
		super.setArgument32(memory)
		super.setAssembler(ASSEMBLER_INSTRUCTION_EXPRESSION(this.sdestination, this.ssource))
	}

	exec ({ context, memory }) {
		// TODO: cast to  `short` ->> (short) context[REGISTER_VALUE_NAME_MAPPER[this.source]]
		memory[Math.floor(Instruction.fix(context[REGISTER_VALUE_NAME_MAPPER[this.destination]]  + this.argument) / 2)] =
		  context[REGISTER_VALUE_NAME_MAPPER[this.source]]

		context.pc  += 6

		return {
			address: Instruction.fix(context[REGISTER_VALUE_NAME_MAPPER[this.destination]]  + this.argument),
			content: context[REGISTER_VALUE_NAME_MAPPER[this.source]]
		}
	}
}
